[papataco14] iP

papataco14 opened the pull request:
# Chad Bot
> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes)) 

Chat Bot helps you keep track of your LIFE! It's, 
- text-based
- easy to learn 
- ~~FAST~~ *SUPER* FAST to use

All you need to do is, 
1. download it from [here](https://nus-cs2103-ay2324s1.github.io/website/schedule/week4/project.html). 
2. double-click it. 
3. add your tasks. 
4. let it manage your tasks for you 😋

And it is ***FREE!***

Features: 
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method: 
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

papataco14 committed:
Rename, Greet, Exit

- Gave chatbot a new name
- Implement an initial skeletal version that greets and exits

papataco14 committed:
Echo

- refactored code
- Added io & chatengine packages
- Added IOHandler interface and ConsoleIO class
- Added ChatEngine class

papataco14 committed:
Add, List

- added ability to store texts by user
- users can retrieve list of text inputs stored

papataco14 committed:
Mark as Done

- added option to mark and unmark tasks
- added Task class to store task info
- added TaskList class to store list of tasks

papataco14 committed:
ToDos, Events, Deadlines

- added support for tracking ToDos, Events & Deadlines
- changed Task to abstract class

papataco14 committed:
Automated Text UI Testing

- Added test cases for testing of basic event adding and querying

papataco14 committed:
Added error handling

- refactored code to allow for better error handling
- added three types of exceptions common for chadbot: invalid commands, arguments, and format

papataco14 committed:
Add delete operation

The task manager does not offer delete task functionality.

Users might want to delete tasks they have completed already or for any other reason.

Let's remove tasks stored in the taskList class and offer the user commands to delete tasks by their index shown to the user.

papataco14 committed:
Refactor ChatEngine command handling and use enums

Hardcoding of commands may lead to more problems in the future.

Handling all command logic in processInput is not maintainable.

Lets use enums for the commands that Chad can accept for improved Readability and Type Safety.

By abstracting out the logic for handling each of the commands into separate functions, commandHandler will now only need to be concerned with deciding which handler function to choose, leading to better separation of concerns.

papataco14 committed:
Implement Save feature

Users couldn't save their list of tasks

Let's allow them to save their current tasks and retrieve it when they load the app again.

papataco14 committed:
Add date time parsing

added data time parsing and storage for deadlines and event start and end.

shuenj commented on src/main/java/chatengine/ChatEngine.java:
```diff
@@ -0,0 +1,168 @@
+
+        try {
+            switch (command) {
+                case MARK:
```
For this module, please note there is no indentation for case clauses.

shuenj commented on src/main/java/chatengine/ChadException.java:
```diff
@@ -0,0 +1,25 @@
+        super(message);
+    }
+
+    public static class InvalidCommandException extends ChadException {
```
I like how you have different exceptions that extend ChadException.

shuenj commented on src/main/java/chatengine/ChatEngine.java:
```diff
@@ -0,0 +1,168 @@
+
+    private void handleEvent(String[] parts) throws ChadException {
+        if (parts.length < 2 || !parts[1].contains(" /from ") || !parts[1].contains(" /to ")) {
+            throw new ChadException.InvalidArgumentException("Invalid format for Event. Use: event {task} /from {start} /to {end}");
```
Perhaps you can consider using line wrapping if the line length exceeds the limit.

shuenj commented on src/main/java/chatengine/Storage.java:
```diff
@@ -0,0 +1,63 @@
+package chatengine;
+
```
I like the ordering of import statements, they are consistent.

shuenj reviewed (commented):
LGTM, just some nits to fix.

awhb commented on src/main/java/chatengine/ChatEngine.java:
```diff
@@ -0,0 +1,168 @@
+    private final Storage storage   ;
+
+    public enum CommandType {
+        BYE, MARK, UNMARK, LIST, TODO, DEADLINE, EVENT, DELETE, UNKNOWN
```
Perhaps you might consider abstracting CommandType to another Java class file?

awhb commented on src/main/java/io/IOHandler.java:
```diff
@@ -0,0 +1,8 @@
+package io;
+
+public interface IOHandler {
+    String readInput();
+    void writeOutput(String output);
+    void greet();
+    void sayGoodbye();
+}
```
Interesting choice to implement an interface for your ConsoleIO class! But since there aren't likely to be other classes implementing IOHandler interface, you might consider not including this interface for simplicity reasons.

awhb reviewed (commented):
Great! 👍 
I'm just nitting at this point.

papataco14 committed:
More OOP

Add more classes to follow OOP paradigmn

papataco14 committed:
Add packages

Added packages

papataco14 committed:
Merge branch 'add-gradle-support'

papataco14 committed:
Config Junit

Configured and added junit test cases

papataco14 committed:
Add javadocs

added javadocs for all non-public methods/classes

papataco14 committed:
Add find feature

allow users to find tasks by keyword

papataco14 committed:
Follow Java Coding Standard

fixed indentation for switch statements

papataco14 committed:
Merge branch 'branch-A-CodingStandard'

papataco14 committed:
Merge branch 'branch-Level-9'

papataco14 committed:
fixed formatting

papataco14 committed:
Bug fixes

fixed bugs

ruo-x commented on src/main/java/chad/chatengine/ChadException.java:
```diff
@@ -0,0 +1,57 @@
+public class ChadException extends Exception {
+
+    /**
+     * Constructs a new ChadException with a specific message.
```
If I am not wrong, there should be an empty line that separates the javadoc descriptions and @params?

ruo-x commented on src/main/java/chad/chatengine/ChatEngine.java:
```diff
@@ -0,0 +1,226 @@
+     * 
+     * @param parts the parsed input array containing the task description.
+     */
+    void handleTodo(String[] parts) {
```
Perhaps you can consider renaming `parts` to `parsedInput` for code clarity purposes?

ruo-x commented on src/main/java/chad/chatengine/Parser.java:
```diff
@@ -0,0 +1,84 @@
+     * Enumeration of possible command types.
+     */
+    public enum CommandType {
+        BYE, MARK, UNMARK, LIST, TODO, DEADLINE, EVENT, DELETE, FIND, UNKNOWN
```
I like how you made use of enumerations!

ruo-x commented on src/main/java/chad/task/Task.java:
```diff
@@ -0,0 +1,109 @@
+        String description = parts[2];
+
+        Task task = null;
+        switch (type) {
```
Small thing but I believe there is no indentation for case clauses.

ruo-x commented on src/main/java/chad/task/Task.java:
```diff
@@ -0,0 +1,109 @@
+/**
+ * Represents a Task that the user wants to keep track off.
+ */
+public abstract class Task {
```
I like how you made Task abstract as it should not be instantiated!

ruo-x reviewed (commented):
LGTM, good job :) I have added some small suggestions on areas where I think code clarity can be improved. Overall, perhaps you can consider breaking down some of the classes (for e,g. ChatEngine) for a cleaner code!

papataco14 committed:
Add GUI

papataco14 committed:
Update tests and build.gradle

This commit includes two main changes:

1. Update unit tests for the `handleEvent` method in the ChatEngine class.
   - Tests now check for the expected output string when a valid event is added.
   - This ensures that the handler method is working as expected and producing the correct output.

2. Modify the `build.gradle` file to set the new entry point for the application.
   - The `mainClass` attribute in the `application` block has been updated to point to the new GUI launcher class.
   - This change is necessary due to the implementation of the new GUI.

These changes collectively improve the robustness of the application and prepare it for the new GUI implementation.

papataco14 committed:
Add assert statements

This commit introduces assert statements in the Parser.java and Storage.java classes to enforce important assumptions and improve code robustness.

papataco14 committed:
Merge pull request #2 from papataco14/branch-A-Assertions

Add assert statements

papataco14 committed:
Improve ChatEngine class code quality

Removed redundant messages and renamed response variable to responseToUser for better clarity

papataco14 committed:
Merge pull request #3 from papataco14/branch-A-CodeQuality

Improve ChatEngine class code quality

papataco14 committed:
Implement task sorting by date

Current TaskList class stores tasks but lacks a feature to sort them
by their deadlines or event dates.

Sorting tasks by date is useful for users who want to see their tasks
in order of urgency without having to manually reorder them.

Implement a method getSortedTasksByDate to sort tasks by their
deadlines and event dates. The method returns a new TaskList
containing the sorted tasks.

We create a sorted copy of the original task list
and encapsulate it in a new TaskList object to avoid modifying the
original list.

This design choice allows us to present a sorted view to the
user while keeping the underlying data unchanged.

Refer to the Java Collections.sort method documentation for
details on how sorting is implemented.

papataco14 committed:
Add Ui and update readme

Updated readme to serve as a user guide and added representative ui image to docs folder

papataco14 committed:
Rename Ui.png

To be detected by bot
